### PR TITLE
Package as a Nix flake to fix problems with dependencies

### PR DESCRIPTION
**Pull request by numinit**
I noticed some complaining in the README about dependency issues... so I solved the problem. :-)

Enjoy.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,9 +9,17 @@ If you are using any of my projects and find them helpful, please consider [dona
...
+
+- Running: `nix run`
+- Building: `nix build`
+    - Wrapper script is located at `./result/bin/m4b-tool`
+- Developing: `nix develop`
+    - When done updating dependencies, run `composer2nix` to update the .nix files
+
 ## Announcement
 
-A few months ago I noticed, that using `PHP` as programming language is a limiting factor for `m4b-tool` and its further development. Multiple dependencies (`php`, `ffmpeg`, `mp4v2`, `fdkaac` and others) make it quite hard to improve the user experience. So I started an experiment, that now has reached an early alpha level and can be tried out. The command line tool is written in `C#`, fully open source and is called `tone`. It already has a pretty decent feature set, so if you would like to try it, here it is:
```
Comment (numinit) [inline README.md:14]: No longer a limitation. :-)

File: default.nix [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+      "--enable-nonfree"
+    ];
+    buildInputs = prev.buildInputs ++ [
+      fdk_aac
+    ];
+  }) else ffmpeg_5-headless;
+in
+m4bToolComposer.overrideAttrs (prev: rec {
+  pname = "m4b-tool";
+  version = "0.5";
```
Comment (numinit) [inline default.nix:39]: Wasn't sure what to call this since it's pulling from master.

File: composer.json [hunk 3]
```diff
@@ -1,4 +1,5 @@
 {
+    "name": "sandreas/m4b-tool",
     "minimum-stability": "dev",
     "prefer-stable": true,
     "platform":{"php":"7.4"},
```
Comment (numinit) [inline composer.json:5]: nixpkgs runs it with 8.2; this caused a couple issues which I solved. There may be more though.

File: default.nix [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      if ! grep -q 'The Nine Situations' audiobook.chapters.txt; then
+        exit 1
+      fi
+
+      if [ ! -f '006-11 The Nine Situations.m4b' ]; then
+        exit 1
+      fi
+    )
+    rm -rf audiobook
+  '';
```
Comment (numinit) [inline default.nix:98]: Note the installCheckPhase - this is a smoketest for the entire installation.

**Comment by sandreas** [comment]
Phew, I always wanted to try nix :-) Well, It may take some weeks until I can integrate this PR, I'm on vacation right now and promised everyone not to be too much into my projects ;)

But thank you very much for providing this, I really appreciate it.

**Comment by numinit** [comment]
No worries, the project was extremely well suited for doing this. Multiple dependencies with the need to bundle them without something like Docker. Still need Nix, but at least the derivations can be packaged to other things like Docker...

**Comment by sandreas** [comment]
@numinit Thanks for your sponsorship. As I said, ATM I'm on vacation, but as soon as I'm back, I'm gonna prioritize issues and put your name (and if you wish your logo) into the project readme.

In the next weeks I plan to publish a sponsors overview on my homepage (https://pilabor.com) with more details about the projects I plan to support. Probably the following projects will be on the list and a survey will be published, because they are the most valuable and transparent projects with charitable purpose available in Germany (where I live):

- WWF (https://www.wwf.de/ - German, https://www.worldwildlife.org/ - International)
- Welthungerhilfe (https://www.welthungerhilfe.org/about-us)
- Ärzte ohne Grenzen (https://www.aerzte-ohne-grenzen.de/work-with-us)

Since charitable donations are only useful if a certain amount is donated (at least 150$), I have to wait until the sponsorships pass this value, but I publish the donation receipt, as soon as it's done.

Thank you for all your support.

**Review by sandreas — COMMENTED** [review]
So first I wanna thank you for submitting this huge PR. I added a few minor questions and comments and would love to hear your comments about these. 

I also added a tag `v0.5-prerelease` (see my comments)

For now, there is nothing that really prevents the merge - I just wanna shortly discuss these things and not force you doing even more work before merging. Feel free to just give short answers and after these I will merge the PR as is or, if you like my suggestions wait for some minor changes.

File: src/library/Command/SplitCommand.php [hunk 5]
```diff
@@ -336,7 +336,9 @@ private function splitChapters()
             $templateParameters = (array)$tag;
             // strip reserved chars from parameters for filename generation (fix #228)
             foreach($templateParameters as $key => $value) {
-                $templateParameters[$key] = static::replaceDirReservedChars($value);
+                if (is_string($value)) {
```
[the remark this answers is not in the document]
Comment (numinit) [reply src/library/Command/SplitCommand.php:339]: An array is passed to strtr, which breaks.

File: composer-env.nix [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                  print($packagesStr);
+              }
+          }
+          else
+              print("[]");
+          ?>
+        '';
+      };
+
+      constructBin = writeTextFile {
```
[the remark this answers is not in the document]
Comment (numinit) [reply composer-env.nix:81]: This file is completely autogenerated by composer2nix. I wouldn't worry about it too much - it's just boilerplate for the Nix composer install to work.

File: default.nix [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  postFixup = ''
+    # Wrap it
+    rm -rf $out/bin
+    mkdir -p $out/bin
+
+    # makeWrapper fails for this on macOS
+    cat >$out/bin/m4b-tool <<EOF
+    #!${runtimeShell}
+    export PATH=${lib.makeBinPath buildInputs}
+    export M4B_TOOL_DISABLE_TONE=true
```
[the remark this answers is not in the document]
Comment (numinit) [reply default.nix:63]: We'd just have to nix package `tone` to get its benefits then.

File: default.nix (hunk 4 above)
Comment (numinit) [reply default.nix:98]: Yeah, that would actually really help. Then we could pull it in the installCheckPhase.

File: default.nix (hunk 2 above)
Comment (numinit) [reply default.nix:39]: And as far as the tag goes, I would make the tag names immutable (0.5-prerelease-$DATE) or something would work. Right now I'm locking it to a particular commit which is less than ideal.

File: composer-env.nix [hunk 8]
```diff
@@ -0,0 +1,244 @@
+# This file originates from composer2nix
```
Comment (numinit) [inline composer-env.nix:1]: See also this - you don't have to worry about this file. `composer2nix` will write out the "correct" version.

File: README.md [hunk 9]
```diff
@@ -9,9 +9,20 @@ If you are using any of my projects and find them helpful, please consider [dona
 <a href="https://github.com/sponsors/sandreas"><img src="./assets/help.svg" width="300" alt="sponsor me and donate" style="margin:auto;"></a>
 </p>
 
+## Usage with Nix
+
+Get [Nix](https://nixos.org/download.html) and ensure that [Flakes](https://nixos.wiki/wiki/Flakes#Permanent) are enabled.
+
+- Running: `nix run github:sandreas/m4b-tool` or `nix run github:sandreas/m4b-tool#m4b-tool-libfdk`
```
Comment (numinit) [inline README.md:16]: BTW you can test:

`nix run github:numinit/m4b-tool`

**Pull request merged**